kernel/time/src/lib.rs

```diff
@@ -0,0 +1,284 @@
+
+impl core::ops::Sub<Instant> for Instant {
+    type Output = Duration;
+
+    fn sub(self, rhs: Instant) -> Self::Output {
+        self.duration_since(rhs)
+    }
+}
+
+/// Register the clock source that can be used to sleep when interrupts are
+/// disabled.
+///
+/// The current early sleeper will only be overwritten by `T` if `frequency` is
+/// larger than the frequency of the current early sleeper.
+///
+/// Returns whether the early sleeper was overwritten.
+pub fn register_early_sleeper<T>(frequency: u64) -> bool
+where
+    T: EarlySleeper,
+{
+    let old_frequency = EARLY_SLEEPER_FREQUENCY.load(Ordering::SeqCst);
+    if frequency > old_frequency {
+        EARLY_SLEEP_FUNCTION.store(T::sleep);
+        EARLY_SLEEPER_FREQUENCY.store(frequency, Ordering::SeqCst);
+        true
+    } else {
+        false
+    }
+}
+
+/// Wait for the given `duration`.
+///
+/// This function spins the current task rather than sleeping it and so, when
+/// possible, the `sleep` crate should be used.
+///
+/// However, unlike the `sleep` crate, this function doesn't rely on interrupts,
+/// and can be used prior to the scheduler being initiated.
+pub fn early_sleep(duration: Duration) {
+    let f = EARLY_SLEEP_FUNCTION.load();
+    f(duration)
+}
+
+/// Register a clock source.
+///
+/// The provided source will overwrite the previous source only if `frequency`
+/// is larger than that of the previous source.
+///
+/// Returns whether the previous source was overwritten.
+pub fn register_clock_source<T>(frequency: u64) -> bool
+where
+    T: ClockSource,
+{
+    let old_frequency = T::ClockType::frequency_atomic().load(Ordering::SeqCst);
+    if frequency > old_frequency {
+        let now_fn = T::ClockType::now_fn();
+        now_fn.store(T::now);
+
+        T::ClockType::store_unit_to_duration_func(T::unit_to_duration);
+        T::ClockType::store_duration_to_unit_func(T::duration_to_unit);
+
+        T::ClockType::frequency_atomic().store(frequency, Ordering::SeqCst);
+
+        true
+    } else {
+        false
+    }
+}
+
+/// Returns the current time.
+///
+/// Monotonic clocks return an [`Instant`] whereas realtime clocks return a
+/// [`Duration`] signifying the time since 12:00am January 1st 1970 (i.e. Unix
+/// time).
+///
+/// # Panics
+///
+/// This function will panic if called prior to registering a clock using
+/// [`register_clock_source`]. [`register_clock_source`] must return [`Ok`] and
+/// the [`ClockType`] of the registered clock must be the same as `T`.
+pub fn now<T>() -> T::Unit
+where
+    T: ClockType,
+{
+    let f = T::now_fn().load();
+    f()
+}
+
+/// A clock source.
+pub trait ClockSource {
+    /// The type of clock (either [`Monotonic`] or [`Realtime`]).
+    type ClockType: ClockType;
+
+    /// The current time according to the clock.
+    ///
+    /// For monotonic clocks this is usually the time since boot, and for
+    /// realtime clocks it's the time since 12:00am January 1st 1970 (i.e.
+    /// Unix time).
```

Comment:
This is good, but it's be even better to be concrete and further clarify that
* monotonic clocks return the time since system boot expressed as an `Instant`, whereas
* wall time clocks return a `Duration` representing the elapsed time since the Unix epoch.

Reply:
there is no way to get the time since boot from an `Instant`, so from the perspective of the API consumer, it just represents an arbitrary point in time.

Reply:
right, that is true, thanks for pointing that out.